heartbeat/skima/schema_validation_results.go

```diff
@@ -0,0 +1,56 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package skima
+
+// SchemaValidationResults the results of executing a schema.
+// They are a flattened map (using dotted paths) of all the values []ValueResult representing the results
+// of the IsDefs.
+type SchemaValidationResults map[string][]ValueResult
+
+func (r SchemaValidationResults) recordResult(path string, result ValueResult) {
+	if r[path] == nil {
+		r[path] = []ValueResult{result}
+	} else {
+		r[path] = append(r[path], result)
+	}
+}
+
+func (r SchemaValidationResults) eachResult(f func(string, ValueResult)) {
+	for path, pathResults := range r {
+		for _, result := range pathResults {
+			f(path, result)
+		}
+	}
+}
+
+// Errors returns a new SchemaValidationResults object consisting only of error data.
+func (r SchemaValidationResults) Errors() SchemaValidationResults {
+	errors := SchemaValidationResults{}
+	r.eachResult(func(path string, vr ValueResult) {
+		if !vr.valid {
+			errors.recordResult(path, vr)
+		}
+	})
+	return errors
+}
+
+// Valid returns true if there are no errors.
+func (r SchemaValidationResults) Valid() bool {
+	// TODO: this is a pretty slow way to do this.
```

Comment:
As this is only for testing should not matter?

Reply:
Yep! I agree. N is pretty small.

One thing @urso did mention is we could eventually use this skima lib as the backend to a user-definable validator in YAML. I think it's important to call out slow functions so that if people use them in the future they're aware they may need optimization.